ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/ApplyMigrationCommand.java

```diff
@@ -56,7 +77,188 @@
+    try {
+      config = MigrationConfig.load(configFile);
+    } catch (KsqlException | MigrationException e) {
+      LOGGER.error(e.getMessage());
+      return 1;
+    }
+
+    return command(
+        config,
+        MigrationsUtil::getKsqlClient,
+        getMigrationsDirFromConfigFile(configFile),
+        Clock.systemDefaultZone()
+    );
+  }
+
+  @VisibleForTesting
+  int command(
+      final MigrationConfig config,
+      final Function<MigrationConfig, Client> clientSupplier,
+      final String migrationsDir,
+      final Clock clock
+  ) {
+    final Client ksqlClient;
+    try {
+      ksqlClient = clientSupplier.apply(config);
+    } catch (MigrationException e) {
+      LOGGER.error(e.getMessage());
+      return 1;
+    }
+
+    if (ValidateMigrationsCommand.validate(config, migrationsDir, ksqlClient)
+        && apply(config, ksqlClient, migrationsDir, clock)) {
+      ksqlClient.close();
+      return 0;
+    } else {
+      ksqlClient.close();
+      return 1;
+    }
+  }
+
+  private boolean apply(
+      final MigrationConfig config,
+      final Client ksqlClient,
+      final String migrationsDir,
+      final Clock clock
+  ) {
+    final int start = getStartVersion(config, ksqlClient);
+    final List<Migration> migrations;
+    LOGGER.info("Loading migration files");
+
+    if (next) {
+      migrations = getAllMigrations(start, start, migrationsDir);
+    } else if (version >= start) {
+      migrations = getAllMigrations(start, version, migrationsDir);;
+    } else if (version > 0 && version < start) {
+      LOGGER.error("Specified version is lower than latest migrated version");
+      return false;
+    } else {
+      migrations = getAllMigrations(start, migrationsDir);
+    }
+
+    for (Migration migration : migrations) {
+      if (!applyMigration(config, ksqlClient, migration, clock)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean applyMigration(
+      final MigrationConfig config,
+      final Client ksqlClient,
+      final Migration migration,
+      final Clock clock
+  ) {
+    LOGGER.info("Applying " + migration.getName() + " version " + migration.getVersion());
+    LOGGER.info(migration.getCommand());
+
+    if (dryRun) {
+      return true;
+    }
+
+    if (!verifyMigrated(config, ksqlClient, migration.getPrevious(), MAX_RETRIES)) {
+      LOGGER.error("Failed to verify status of version " + migration.getPrevious());
+      return false;
+    }
+
+    final String executionStart = Long.toString(clock.millis());
+
+    if (
+        !updateState(config, ksqlClient, MigrationState.RUNNING, executionStart, migration, clock)
+    ) {
+      return false;
+    }
+
+    try {
+      final List<String> commands = Arrays.stream(migration.getCommand().split(";"))
+          .filter(s -> s.length() > 1)
+          .collect(Collectors.toList());
+      for (final String command : commands) {
+        ksqlClient.executeStatement(command + ";").get();
```

Comment:
This doesn't work for INSERT VALUES statements. We can handle this case in a follow-up PR, but INSERT VALUES statements need to be submitted via the `insertInto()` method rather than `executeStatement()`.

Reply:
It just occurred to me that translating `INSERT VALUES` statements into `insertInto()` calls is actually really annoying. Let me think some more about the best way to handle this. We might want to make another minor update to the client.